File: frontend/src/services/cvrp.ts
const isDev = import.meta.env.DEV

const API_BASE_URL = isDev ? 'http://localhost:8000/api/v1/cvrp' : '/api/v1/cvrp'

export interface CVRPRequest {
  waste_type: string
  n_vehicles: number
  vehicle_capacity: number
  max_runtime: number
  waste_per_centroid: number
  load_unit: string
  edge_modifications: Array<{ u: number; v: number; action: string; speed_kph?: number }>
}

export interface CVRPRouteSegment {
  route_id: number
  trip_id: number
  path_coordinates: [number, number][]
  load_kg: number
}

export interface CVRPEdgeLoad {
  u: number
  v: number
  load: number
}

export interface CVRPSolveResponse {
  n_routes: number
  n_missing_clients: number
  total_distance_m: number
  route_segments: CVRPRouteSegment[]
  edge_loads: CVRPEdgeLoad[]
  load_unit: string
  solve_time_ms: number
  centroids_used: number
}

export async function solveCVRP(req: CVRPRequest): Promise<CVRPSolveResponse> {
  const response = await fetch(`${API_BASE_URL}/solve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req)
  })
  if (!response.ok) {
    const detail = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(detail.detail ?? 'CVRP solve failed')
  }
  return response.json()
}

export async function fetchCVRPCentroids(wasteType: string): Promise<GeoJSON.FeatureCollection> {
  const response = await fetch(`${API_BASE_URL}/centroids?waste_type=${wasteType}`)
  if (!response.ok) {
    throw new Error(`Failed to fetch centroids for ${wasteType}`)
  }
  return response.json()
}

File: backend/app/models/cvrp.py
"""CVRP models for API requests and responses."""

from typing import List

from pydantic import BaseModel, Field

from app.models.route import EdgeModification


class CVRPRequest(BaseModel):
    """Request body for CVRP solve endpoint."""

    waste_type: str = Field(default="DI", description="Waste type: DI, DV, PC, or VE")
    n_vehicles: int = Field(default=5, ge=1, le=50, description="Number of collection vehicles")
    vehicle_capacity: int = Field(default=5000, ge=100, description="Vehicle capacity in kg")
    max_runtime: int = Field(default=10, ge=1, le=120, description="Solver max runtime in seconds")
    waste_per_centroid: int = Field(default=10, ge=1, description="Waste per centroid in kg")
    load_unit: str = Field(default="kg", description="Load unit: 'kg' or 'kg_m'")
    edge_modifications: List[EdgeModification] = Field(
        default_factory=list, description="Edge modifications to respect during routing"
    )


class CVRPRouteSegment(BaseModel):
    """A single routed path segment for one vehicle trip."""

    route_id: int
    trip_id: int
    path_coordinates: List[List[float]]  # [[lon, lat], ...]
    load_kg: float


class CVRPEdgeLoad(BaseModel):
    """Aggregated load passing through a single graph edge."""

    u: int
    v: int
    load: float


class CVRPSolveResponse(BaseModel):
    """Response from CVRP solve endpoint."""

    n_routes: int
    n_missing_clients: int
    total_distance_m: float
    route_segments: List[CVRPRouteSegment]
    edge_loads: List[CVRPEdgeLoad]
    load_unit: str
    solve_time_ms: float
    centroids_used: int
